mlir/lib/Transforms/Utils/DialectConversion.cpp

```diff
@@ -1620,6 +1620,8 @@ void ConversionPatternRewriterImpl::notifyMatchFailure(
     if (config.notifyCallback)
       config.notifyCallback(diag);
   });
+  if (config.listener)
+    config.listener->notifyMatchFailure(loc, reasonCallback);
```

Comment:
matchFailure are meant to be eliminated entirely from a release build, as such the entire body of this method is intentionally behind LLVM_DEBUG.

Reply:
I have 3 reasons to disagree.

1. This is already done in greedy rewriter.
https://github.com/llvm/llvm-project/blob/ffec63b6a54f719976443cc17b6226030be17c52/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp#L760-L769

2. I disagree that match failure should be eliminated from a release build. For example, in my use case, we use `notifyMatchFailure` on patterns that we _must_ run to support conversions that lower from higher-level to lower-level dialects. I consider a failure to match a required lowering pattern as something more helpful to the user than just emitting "failure to legalize operation explicitly marked illegal".

3. Even if we only notify match failure in a non-release build, my understanding of the `LLVM_DEBUG` macro is that it only runs if the command line option `--debug` or `--debug-only=<debug type>" is specified.

Reply:
If this is not an intended purpose for `notifyMatchFailure`, then can you help me explain why in the documentation?
I believe myself and others who agree with me probably just thought that match failure could be used as a way to signal a specific reason for failure, so I thought it made sense to add a way to use a listener to forward those.

Reply:
>  then can you help me explain why in the documentation?

It is meant as a zero-cost (in production) debugging feature, like assertions or every other "-debug" enabled messages. If you feel this needs a clarification in the documentation, we can do this of course.

The greedyRewriterAddition smells wrong to me, I'll follow up with @matthias-springer on this.

The current setup for patterns is that failure to match isn't a problem and is expected: there may be other patterns that will succeed. So there messages aren't meant to be useful to any end users right now.


Note that we could revisit the design, I'm pointing out what the consensus was in the core team when we designed the feature.

Reply:
> This is already done in greedy rewriter, as added by @matthias-springer (who seems to agree with me on the goals of this PR):

I have to admit I didn't think much about `notifyMatchFailure` when I added listener support to the greedy pattern rewrite driver. I just forwarded all notifications to the listener.

> I disagree that match failure should be eliminated from a release build. For example, in my use case, we use notifyMatchFailure on patterns that we must run to support conversions that lower from higher-level to lower-level dialects. I consider a failure to match a required lowering pattern as something more helpful to the user than just emitting "failure to legalize operation explicitly marked illegal".

So the use case here is to produce better error messages. What happens after a "required pattern" reports a match failure? Sending the error message via the listener seems a bit convoluted. Could you directly `op->emitError()` and `return failure()` in the pattern?